File: src/RmSQL.c
#include <stdio.h>
#include <sys/types.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <msql.h>

m_result *result;
m_field *field;
char *errMsg;

void RgetErrMsg(msg)
	char **msg;
{
	msg[0] = errMsg;
}

void RmsqlConnect(host, stat)
	char **host;
	int *stat;
{
	stat[0] = msqlConnect(host[0]);
	if (stat[0] < 0) errMsg = msqlErrMsg; 	
}

void RmsqlSelectDB(sock, dbName, stat)
	int *sock;
	char **dbName;
	int *stat;
{
	stat[0] = msqlSelectDB(sock[0], dbName[0]);
	if (stat[0] < 0) errMsg = msqlErrMsg; 	
}

void RmsqlQuery(sock, query, stat)
	int *sock;
	char **query;
	int *stat;
{
	stat[0] = msqlQuery(sock[0], query[0]);
	if (stat[0] < 0) errMsg = msqlErrMsg; 	
}

void RmsqlStoreResult()
{
	result = msqlStoreResult();
}

void RmsqlFetchRow(data, num, stat)
	char **data;
	int *num;
	int *stat;
{
	m_row cur;
	int i;
	 
	if (!result)
	{
		errMsg = "No results available";
		stat[0] = -1;
		return;
	}
	if (cur = msqlFetchRow(result))
	{
		for (i=0; i< num[0]; i++)
			data[i] = cur[i];
	} else {
		errMsg ="No more data";
		stat[0] = -1;
	}	
}


void RmsqlFreeResult()
{
	if (!result) return;
	msqlFreeResult(result);
}

void RmsqlDataSeek(pos, stat)
	int *pos, *stat;
{
	if (!result)
	{
		errMsg = "No results available";
		stat[0] = -1;
		return;
	}
	msqlDataSeek(result, pos[0]);
}

void RmsqlNumRows(num, stat)
	int *num, *stat;
{
	if (!result)
	{
		errMsg = "No results available";
		stat[0] = -1;
		return;
	}
	num[0] = msqlNumRows(result);
}

void RmsqlFetchField(attr, type, stat)
	char **attr;
	int *type, *stat;
{
	if (!result)
	{
		errMsg = "No results available";
		stat[0] = -1;
		return;
	}
	if (field = msqlFetchField(result))
	{
		attr[0] = field->name;
		attr[1] = field->table;
		type[0] = field->type;
	} else {
		errMsg ="No more data";
		stat[0] = -1;
	}
}

void RmsqlFieldSeek(pos, stat)
	int *pos, *stat;
{
	if (!result)
	{
		errMsg = "No results available";
		stat[0] = -1;
		return;
	}
	msqlFieldSeek(result, pos[0]);
}

void RmsqlNumFields(num, stat)
	int *num, *stat;
{
	if (!result)
	{
		errMsg = "No results available";
		stat[0] = -1;
		return;
	}
	num[0] = msqlNumFields(result);
}	

void RmsqlListDBs(sock)
	int *sock;
{
	result = msqlListDBs(sock[0]);
}

void RmsqlListTables(sock)
	int *sock;
{
	result = msqlListTables(sock[0]);
}

void RmsqlListFields(sock, tableName)
	int *sock;
	char **tableName;
{
	result = msqlListFields(sock[0], tableName[0]);
}

void RmsqlListIndex(sock, tableName, index)
	int *sock;
	char **tableName, **index;
{
	result = msqlListIndex(sock[0], tableName[0], index[0]);
}

void RmsqlClose(sock)
	int *sock;
{
	msqlClose(sock[0]);
}

File: R/RmSQL.R
.First.lib <- function(lib, pkg)
{
	library.dynam(pkg, lib.loc = lib);
}

msqlGetErrMsg <- function()
{
	erg <- .C("RgetErrMsg", err = as.character(c("e")));
	return(erg$err);
}

msqlConnect <- function(host)
{
	erg <- .C("RmsqlConnect", host, stat = as.integer(1));
	
	return(erg$stat)
};

msqlSelectDB <- function(sock, dbName)
{
	erg<-.C("RmsqlSelectDB", sock, dbName, stat = as.integer(1));
	return(erg$stat);
};

msqlQuery <- function(sock, query)
{
	erg <- .C("RmsqlQuery", sock, query, stat = as.integer(1));
	return(erg$stat);
}

msqlStoreResult <- function()
{
	erg <- .C("RmsqlStoreResult");
}

msqlFetchRow <- function()
{
	num <- msqlNumFields()$num;
	erg <- .C("RmsqlFetchRow", data = as.character(paste("D", 1:num, sep="")), num, stat = as.integer(1));
	return(erg);
}

msqlFreeResult <- function()
{
	.C("RmsqlFreeResult");
}

msqlDataSeek <- function(pos)
{
	erg <- .C("RmsqlDataSeek", as.integer(pos), stat = as.integer(1));
	return(erg$stat);
}


msqlNumRows <- function()
{
	erg <- .C("RmsqlNumRows", num = as.integer(1), stat = as.integer(1));
	return(erg);
}

msqlFetchField <- function()
{
	erg <- .C("RmsqlFetchField", attr = as.character(c("name","table")), type = as.integer(1), stat = as.integer(1));
	
	#	name: 	attribute name
	#	table: 	name of table
	#	type:	1 	int
	#		2	char
	#		3	double
	#		6	text
	#		7	date

	return(erg);
}

msqlFieldSeek <- function(pos)
{
	erg <- .C("RmsqlFieldSeek", as.integer(pos), stat = as.integer(1));
	return(erg$stat);
}

msqlNumFields <- function()
{
	erg <- .C("RmsqlNumFields", num = as.integer(1), stat = as.integer(1));
	return(erg);
}

msqlListDBs <- function(sock)
{
	erg <- .C("RmsqlListDBs", sock);
}

msqlListTables <- function(sock)
{
	erg <- .C("RmsqlListTables", sock);
}

msqlListFields <- function(sock, tableName)
{
	erg <- .C("RmsqlListFields", sock, tableName);
}

msqlListIndex <- function(sock, tableName, index)
{
	erg <- .C("RmsqlListIndex", sock, tableName, index);
}

msqlClose <- function(sock)
{
	.C("RmsqlClose", sock);
}

